### PR TITLE
fix: don't exit on reload if there is a syntax error

### LINKED ISSUE #213: functions-framework --debug exits if you save with a syntax error
Steps to reproduce:

* Start things up normally
```
functions-framework \
    --source=cloud_function/main.py \
    --target=hello_get \
    --host=localhost \
    --port=18000 \
    --debug
```
<details><summary>This starts up as expected:</summary>
```
 * Serving Flask app 'hello_get' (lazy loading)
 * Environment: production
   WARNING: This is a development server. Do not use it in a production deployment.
   Use a production WSGI server instead.
 * Debug mode: on
WARNING: This is a development server. Do not use it in a production deployment. Use a production WSGI server instead.
 * Running on http://localhost:18000
Press CTRL+C to quit
 * Restarting with watchdog (fsevents)
 * Debugger is active!
 * Debugger PIN: 104-682-644
```
</details>

Make a change that introduces a syntax error, and you get:
```
 * Detected change in '/Users/david/src/opvia/app/scripts_v3/cloud_function/main.py', reloading
 * Restarting with watchdog (fsevents)
Traceback (most recent call last):
  File "/Users/david/src/opvia/app/.venv/bin/functions-framework", line 8, in <module>
    sys.exit(_cli())
  File "/Users/david/src/opvia/app/.venv/lib/python3.10/site-packages/click/core.py", line 1130, in __call__
    return self.main(*args, **kwargs)
  File "/Users/david/src/opvia/app/.venv/lib/python3.10/site-packages/click/core.py", line 1055, in main
    rv = self.invoke(ctx)
  File "/Users/david/src/opvia/app/.venv/lib/python3.10/site-packages/click/core.py", line 1404, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/Users/david/src/opvia/app/.venv/lib/python3.10/site-packages/click/core.py", line 760, in invoke
    return __callback(*args, **kwargs)
  File "/Users/david/src/opvia/app/.venv/lib/python3.10/site-packages/functions_framework/_cli.py", line 37, in _cli
    app = create_app(target, source, signature_type)
  File "/Users/david/src/opvia/app/.venv/lib/python3.10/site-packages/functions_framework/__init__.py", line 289, in create_app
    spec.loader.exec_module(source_module)
  File "<frozen importlib._bootstrap_external>", line 879, in exec_module
  File "<frozen importlib._bootstrap_external>", line 1017, in get_code
  File "<frozen importlib._bootstrap_external>", line 947, in source_to_code
  File "<frozen importlib._bootstrap>", line 241, in _call_with_frames_removed
  File "/Users/david/src/opvia/app/scripts_v3/cloud_function/main.py", line 11
    ///
    ^^
SyntaxError: invalid syntax
```

and the process exits

Potential solution:

Apparently this used to be an issue in Flask, but isn't anymore (see https://stackoverflow.com/a/68445450 ). I did some digging to work out how they fixed it, and I think I have a solution:

I think that functions-framework basically needs to do something like this:

https://github.com/pallets/flask/pull/4718/files#diff-fed8939d1905b99bba605ad91e9ccdf5ede6223e03ccbc3f0121853035051e62R935-R950

in functions_framework/init.py", line 289, in create_app where it does:
```python
    with _app.app_context():
        spec.loader.exec_module(source_module)
```
* catch the error
* check werkzeug.serving.is_running_from_reloader().
  * if true, print the error and create a dummy app that explodes when called
  * else re-raise the error

### PR DESCRIPTION
**Pull request by alsuren**
fixes #213 

Now, if you introduce a syntax error, you immediately get:

```
 * Detected change in '/Users/david/src/opvia/app/scripts_v3/cloud_function/main.py', reloading
 * Restarting with watchdog (fsevents)
Traceback (most recent call last):
  File "/Users/david/src/opvia/app/.venv/lib/python3.10/site-packages/functions_framework/__init__.py", line 290, in create_app
    spec.loader.exec_module(source_module)
  File "<frozen importlib._bootstrap_external>", line 879, in exec_module
  File "<frozen importlib._bootstrap_external>", line 1017, in get_code
  File "<frozen importlib._bootstrap_external>", line 947, in source_to_code
  File "<frozen importlib._bootstrap>", line 241, in _call_with_frames_removed
  File "/Users/david/src/opvia/app/scripts_v3/cloud_function/main.py", line 11
    ///
    ^^
SyntaxError: invalid syntax
 * Debugger is active!
 * Debugger PIN: 104-682-644
```

and then if you curl your fuction endpoint, you get the debug html page for the syntax error, and the backtrace gets printed out to stderr again.

If you attempt to *start* the server when there is already a syntax error present, it correctly exits with the error, as before (the only difference being that the backtrace points at the `raise e from None` line rather than the `spec.loader.exec_module(source_module)` line). 

This behaviour matches the upstream `flask run` behaviour now.

Feel free to be brutal about the quality of my python code: the last time I wrote significant production python, it was before python 2.7 was deprecated ;-)

### CODE REVIEW

**Comment by alsuren** [comment]
We are now using this patch both locally and in our cloud-function environment. We've not run into any issues with it yet.

What are the next steps for this?

**Review by HKWinterhalter — APPROVED** [review]

**Comment by HKWinterhalter** [comment]
It looks like this will need test coverage and linting to pass

**Comment by alsuren** [comment]
`python -m tox -e lint` and `python -m tox -e py-macos-latest` pass locally (coverage is at 100%)

To work around the "This workflow requires approval from a maintainer" message, I made a PR against my own fork (https://github.com/alsuren/functions-framework-python/pull/1). The 7 "Expected" checks seem to pass on that PR, so I'm confident that they should pass on this PR too.

I'm assuming that the other tests are not required, or just not configured correctly on my fork. Shout if you need me to look into any more issues.

**Comment by HKWinterhalter** [comment]
Thank you for fixing this @alsuren !

**Pull request merged**
